esp/esp/program/controllers/autoscheduler/scoring.py

```diff
@@ -0,0 +1,1136 @@
+            for furnishing in room.furnishings.itervalues():
+                if furnishing.name not in furnishing_values:
+                    furnishing_values[furnishing.name] = set()
+                furnishing_values[furnishing.name].add(furnishing.value)
+        self.requests_to_ignore = set()
+        for name, values in furnishing_values.iteritems():
+            if len(values) == 1:
+                self.requests_to_ignore.add(name)
+        self.total_score = 0.0
+        for section in schedule.class_sections.itervalues():
+            if section.is_scheduled():
+                self.total_score += self.process_section(
+                    section, section.assigned_roomslots[0].room)
+
+    @util.timed_func("ResourceValueMatchingScorer")
+    def update_schedule_section(self, section, start_roomslot):
+        """Update the internal state to reflect the scheduling of the specified
+        section to start at the specified roomslot."""
+        self.total_score += self.process_section(section, start_roomslot.room)
+
+    @util.timed_func("ResourceValueMatchingScorer")
+    def update_unschedule_section(self, section):
+        """Update the internal state to reflect the uncsheduling of the
+        specified section."""
+        self.total_score -= self.process_section(
+            section, section.assigned_roomslots[0].room)
+
+
+class RoomConsecutivityScorer(BaseScorer):
+    """Try to schedule classes consecutively in classrooms. This is a good
+    heuristic for being able to schedule long classes later, as well as for
+    minimizing having to repeatedly clean or lock/unlock rooms."""
+
+    def boundary_complete_before(self, roomslot, ignore_section=None):
+        room = roomslot.room
+        start_idx = roomslot.index()
+        if start_idx > 0:
+            prev_roomslot = room.availability[start_idx - 1]
+            return (prev_roomslot.assigned_section is not None
+                    and prev_roomslot.assigned_section != ignore_section) \
+                or not util.contiguous(
+                    prev_roomslot.timeslot, roomslot.timeslot)
+        else:
+            return True
+
+    def boundary_complete_after(self, roomslot, ignore_section=None,
+                                avoid_doublecount=False):
+        next_roomslot = roomslot.next()
+        return next_roomslot is None \
+            or (not avoid_doublecount
+                and next_roomslot.assigned_section is not None
+                and next_roomslot.assigned_section != ignore_section) \
+            or not util.contiguous(
+                roomslot.timeslot, next_roomslot.timeslot)
+
+    @util.timed_func("RoomConsecutivityScorer")
+    def score_schedule(self):
+        """Returns a score in the range [0, 1] for the schedule reflected in its
+        current state."""
+        # Return the fraction of complete boundaries, i.e. start or end of a
+        # section with either no timeslot before/after or another room
```

Comment:
do you mean something like "another section scheduled in the same room" instead of "another room"?

I had to think for a while to understand. Is this understanding true: a boundary is incomplete iff it's between two available and contiguous timeslots, exactly one of which has a section scheduled? Might be worth adding a description like that.